src/zope/interface/adapter.py

```diff
@@ -177,6 +312,44 @@ def registered(self, required, provided, name=u''):
 
         return components.get(name)
 
+    @classmethod
+    def _allKeys(cls, components, i, parent_k=()):
+        if i == 0:
+            for k, v in components.items():
+                yield parent_k + (k,), v
+        else:
+            for k, v in components.items():
+                new_parent_k = parent_k + (k,)
+                for x, y in cls._allKeys(v, i - 1, new_parent_k):
+                    yield x, y
```

Comment:
I can not estimate good how many component.items() there are. But from my gut feeling I guess enough to better first store `cls._allKeys` in a local variable and avoid one load at byte-code level. Like:
```suggestion
        else:
            allKeys = cls._allKeys
            for k, v in components.items():
                new_parent_k = parent_k + (k,)
                for x, y in allKeys(v, i - 1, new_parent_k):
                    yield x, y
```

Reply:
I appreciate your concern for performance!

Here, because of the recursion and use of generators, all of which are implemented in Python and spread across a few different functions, plus the fact that there's lots of tuple creation and unpacking going on, I strongly suspect our time won't be dominated by that attribute lookup. 

But! I haven't benchmarked! I haven't benchmarked because I don't know what a performance goal for this code should be. The only use-case I have for it now is in rebuilding the datastructures with the `rebuild()` method, and that's not particularly performance sensitive (again, in my use-case) — it's a migration-type activity. So I was more concerned about code clarity/maintainability, and secondarily (because of the recursion) trying to keep the stack frames small. 

Perhaps, if need be, we can revisit when someone comes up with different use cases?

Reply:
My  thinking here: It's still readable, it does not hurt, it is fast from the beginning.

Reply:
Given this fixture:
```python
from zope.interface.adapter import AdapterRegistry
from zope.interface import Interface
import pprint

reg = AdapterRegistry()
for i in range(10000):
    iface = type(Interface)(
        'I' + str(i),
        (Interface,),
        {}
    )
    reg.subscribe([object()], iface, i)
```

Looping across `reg.allSubscriptions()` with a simple `for _ in reg.allSubscriptions(): pass` calls `reg._allKeys` 16,221,622 times.

With the code as written, IPython's `%timeit` says this loop takes:
```
15.6 ms ± 142 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Making the change suggested here results in:
```
15.8 ms ± 278 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

So, no appreciable benefit. The bottlenecks are elsewhere.

Reply:
Indeed fast enough.

Reply:
I was testing on CPython 3.8. When I saw that the second version actually got a bit slower, I remembered that starting in Python 3.7 there's the new bytecode `LOAD_METHOD` and `CALL_METHOD` that optimize method calling and avoid creating temporary method objects. So this form of optimization can actually turn out to be a pessimazation on those versions!

Given:
```python
In [1]: class C:
   ...:     def meth(self):
   ...:         pass
   ...:     def call_meth_direct(self):
   ...:         self.meth()
   ...:     def call_meth_indirect(self):
   ...:         m = self.meth
   ...:         m()
   …:
```

The direct call results in 10 bytecodes:
```pycon
In [3]: dis.dis(C.call_meth_direct)
  5           0 LOAD_FAST                0 (self)
              2 LOAD_METHOD              0 (meth)
              4 CALL_METHOD              0
              6 POP_TOP
              8 LOAD_CONST               0 (None)
             10 RETURN_VALUE
```

While the version that stores it in a variable first misses the optimization and is bigger:
```pycon
In [5]: dis.dis(C.call_meth_indirect)
  7           0 LOAD_FAST                0 (self)
              2 LOAD_ATTR                0 (meth)
              4 STORE_FAST               1 (m)

  8           6 LOAD_FAST                1 (m)
              8 CALL_FUNCTION            0
             10 POP_TOP
             12 LOAD_CONST               0 (None)
             14 RETURN_VALUE
```

So what about Python 2.7, where the optimization isn't present? I get 14.7ms as-is, and between 14.3ms and 15.9ms with the modified version on two subsequent runs (IPython on 2.7 doesn't report standard deviation, just the best time).